### PR TITLE
Add support for IBM Cloud Kubernetes Service IKS

### PR DESCRIPTION
**Pull request by nasark**
- added refresh support for IBM Cloud Kubernetes Service

Screen Shot 2021-09-08 at 5 07 47 PM

Depends on:
* [x] [Update 'ibm_cloud_iam' gem to include id_token in token response](https://github.com/IBM-Cloud/ibm-cloud-sdk-ruby/pull/66)
* [x] [Add IKS Icon](https://github.com/ManageIQ/manageiq-decorators/pull/54)

Reference:
- https://github.com/ManageIQ/manageiq-providers-ibm_cloud/issues/149

### CODE REVIEW

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  def self.get_token(api_key)
+    require 'ibm_cloud_iam'
+
+    iam_token_api = IbmCloudIam::TokenOperationsApi.new
+    grant_type = 'urn:ibm:params:oauth:grant-type:apikey'
+    header_params = {
+      'Content-Type'  => 'application/x-www-form-urlencoded',
+      'Authorization' => 'Basic a3ViZTprdWJl',
+      'cache-control' => 'no-cache'
+    }
```
Comment (nasark) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:83]: https://cloud.ibm.com/docs/containers?topic=containers-cs_api_install&locale=en#kube_api

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 2]
```diff
@@ -0,0 +1,201 @@
+class ManageIQ::Providers::IbmCloud::ContainerManager < ManageIQ::Providers::Kubernetes::ContainerManager
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:1]: @nasark can you add `ManageIQ::Providers::Kubernetes::ContainerManager.include(ActsAsStiLeafClass)` above this like https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/container_manager.rb#L1 ?

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  def self.description
+    @description ||= "IBM Cloud Kubernetes Service".freeze
+  end
+
+  def self.display_name(number = 1)
+    n_('Container Provider (IBM Cloud)', 'Container Providers (IBM Cloud)', number)
+  end
+
+  def self.default_port
+    443
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:23]: Is 443 the default port for the API server?

Comment (nasark) [reply app/models/manageiq/providers/ibm_cloud/container_manager.rb:23]: The port that was used on my cluster was `31821`, but I'm not sure if `443` is the default
i.e.
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: <ca-path>
    server: https://c103.ca-tor.containers.cloud.ibm.com:31821
  name: <cluster-name>
```

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      :region => provider_region
+    )
+  rescue IbmCloudIam::ApiError => err
+    error_message = JSON.parse(err.response_body)["message"]
+    _log.error("IAM authentication failed: #{err.code} #{error_message}")
+  end
+
+  def self.kubernetes_auth_options(options)
+    {
+      :bearer_token => options[:bearer]
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:41]: @nasark what you want to do is get the token dynamically in this method, rather than in `connect_options` and `verify_credentials` because this is what is re-run each time we connect in the refresh watch threads.  We save the "connect_options" and re-connect with the same options so you'd be saving a stale bearer token.

recommend:
```
:bearer_token => get_token(options[:api_key]).id_token
```

Then passing in the `:api_key` to options in `connect_options` and `verify_credentials`

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    443
+  end
+
+  def connect_options(options = {})
+    authentication = authentication_best_fit(options.fetch(:auth_type, "bearer"))
+    token = self.class.get_token(authentication.auth_key).id_token
+
+    super.merge(
+      :bearer => token,
+      :region => provider_region
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:32]: Do you actually need the region?  I don't see it passed to `self.class.get_token()`

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  def self.default_port
+    443
+  end
+
+  def connect_options(options = {})
+    authentication = authentication_best_fit(options.fetch(:auth_type, "bearer"))
+    token = self.class.get_token(authentication.auth_key).id_token
+
+    super.merge(
+      :bearer => token,
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:31]: As per below recommend:
```
super.merge(:api_key => authentication.auth_key)
```

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                                         OpenSSL::X509::Certificate.new(certificate_authority)
+                                       )
+    end
+
+    options = {
+      :bearer      => token,
+      :region      => region,
+      :ssl_options => {
+        :verify_ssl => security_protocol == 'ssl-without-validation' ? OpenSSL::SSL::VERIFY_NONE : OpenSSL::SSL::VERIFY_PEER,
+        :cert_store => cert_store
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:67]: Did `:ssl_options => {:certificate_authority => certificate_authority}` not work here?

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    unless certificate_authority.nil?
+      cert_store = OpenSSL::X509::Store.new
+                                       .add_cert(
+                                         OpenSSL::X509::Certificate.new(certificate_authority)
+                                       )
+    end
+
+    options = {
+      :bearer      => token,
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager.rb:63]: Per above should be `:api_key => api_key,`

File: app/models/manageiq/providers/ibm_cloud/container_manager/container.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+class ManageIQ::Providers::IbmCloud::ContainerManager::Container < ManageIQ::Providers::Kubernetes::ContainerManager::Container
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager/container.rb:1]: Same comment as here: https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/272/files#r707588249

File: app/models/manageiq/providers/ibm_cloud/container_manager/container_group.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+class ManageIQ::Providers::IbmCloud::ContainerManager::ContainerGroup < ManageIQ::Providers::Kubernetes::ContainerManager::ContainerGroup
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager/container_group.rb:1]: Same comment as here: https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/272/files#r707588249

File: app/models/manageiq/providers/ibm_cloud/container_manager/container_node.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+class ManageIQ::Providers::IbmCloud::ContainerManager::ContainerNode < ManageIQ::Providers::Kubernetes::ContainerManager::ContainerNode
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager/container_node.rb:1]: Same comment as here: https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/272/files#r707588249

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb (hunk 3 above)
Comment (agrare) [reply app/models/manageiq/providers/ibm_cloud/container_manager.rb:23]: Yeah I don't think that IKS has a default port so best to leave this out and not have a default port in the params_for_create method below

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb (hunk 5 above)
Comment (nasark) [reply app/models/manageiq/providers/ibm_cloud/container_manager.rb:32]: Yeah we don't need it, I missed that thanks

File: app/models/manageiq/providers/ibm_cloud/container_manager.rb (hunk 7 above)
Comment (nasark) [reply app/models/manageiq/providers/ibm_cloud/container_manager.rb:57]: I was unable to get this to work, seems like the cert is dumped as a string:
```
[----] E, [2021-09-13T15:29:08.799616 #41510:93a8] ERROR -- evm: MIQ(MiqQueue#deliver) Message id: [4243], Error: [HTTP status code , SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate in certificate chain)]
[----] E, [2021-09-13T15:29:08.799734 #41510:93a8] ERROR -- evm: [Kubeclient::HttpError]: SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate in certificate chain)  Method:[block (2 levels) in <class:LogProxy>]
[----] E, [2021-09-13T15:29:08.799818 #41510:93a8] ERROR -- evm: /Users/nasarkhan/.rvm/gems/ruby-2.7.3/gems/kubeclient-4.9.2/lib/kubeclient/common.rb:130:in `rescue in handle_exception'
/Users/nasarkhan/.rvm/gems/ruby-2.7.3/gems/kubeclient-4.9.2/lib/kubeclient/common.rb:120:in `handle_exception'
/Users/nasarkhan/.rvm/gems/ruby-2.7.3/gems/kubeclient-4.9.2/lib/kubeclient/common.rb:596:in `fetch_entities'
/Users/nasarkhan/.rvm/gems/ruby-2.7.3/gems/kubeclient-4.9.2/lib/kubeclient/common.rb:583:in `load_entities'
/Users/nasarkhan/.rvm/gems/ruby-2.7.3/gems/kubeclient-4.9.2/lib/kubeclient/common.rb:134:in `discover'
/Users/nasarkhan/Documents/manageiq/plugins/manageiq-providers-kubernetes/app/models/manageiq/providers/kubernetes/container_manager.rb:850:in `kubernetes_connect'
/Users/nasarkhan/Documents/manageiq/plugins/manageiq-providers-kubernetes/app/models/manageiq/providers/kubernetes/container_manager.rb:64:in `raw_connect'
/Users/nasarkhan/Documents/manageiq/plugins/manageiq-providers-ibm_cloud/app/models/manageiq/providers/ibm_cloud/container_manager.rb:58:in `verify_credentials'
```

I found the `cert_store` that I've employed over here: https://github.com/abonas/kubeclient#ssl

**Comment by nasark** [comment]
@agrare Thanks for the feedback, I've made the changes

File: app/models/manageiq/providers/ibm_cloud/container_manager/container.rb [hunk 12]
```diff
@@ -0,0 +1,4 @@
+ManageIQ::Providers::Kubernetes::ContainerManager.include(ActsAsStiLeafClass)
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager/container.rb:1]: Ah sorry I didn't fully explain this.

So if we subclass a model (aka it has a backing database table, not just a ruby class) from another plugin, we have to declare that we are doing this so that STI queries don't pull back both (in this case) `Kubernetes::ContainerManager::Container` _and_ `IbmCloud::ContainerManager::Container` types.

We do this by including the `ActsAsStiLeafClass` into the class we're inheriting from.

Long story short, this should be `ManageIQ::Providers::Kubernetes::ContainerManager::Container.include(ActsAsStiLeafClass)`

Comment (nasark) [reply app/models/manageiq/providers/ibm_cloud/container_manager/container.rb]: Makes sense, thanks for the explanation

File: config/secrets.defaults.yml [hunk 13]
```diff
@@ -14,3 +14,9 @@ test:
     secret_key: IBM_CLOUD_OBJECT_STORAGE_SECRET_KEY
   ibm_cloud_object_storage:
     <<: *ibm_cloud_object_storage_defaults
+  iks_defaults: &iks_defaults
+    api_key: IBM_CLOUD_IKS_API_KEY
+    hostname: IBM_CLOUD_IKS_HOSTNAME
+    port: IBM_CLOUD_IKS_PORT
```
Comment (agrare) [inline config/secrets.defaults.yml:20]: @nasark `port` has to be an integer: https://app.travis-ci.com/github/ManageIQ/manageiq-providers-ibm_cloud/jobs/537165737#L2152-L2178
It can be anything, `1234` for example

File: config/secrets.defaults.yml [hunk 14]
```diff
@@ -14,3 +14,9 @@ test:
     secret_key: IBM_CLOUD_OBJECT_STORAGE_SECRET_KEY
   ibm_cloud_object_storage:
     <<: *ibm_cloud_object_storage_defaults
+  iks_defaults: &iks_defaults
+    api_key: IBM_CLOUD_IKS_API_KEY
+    hostname: IBM_CLOUD_IKS_HOSTNAME
```
Comment (agrare) [inline config/secrets.defaults.yml:19]: Also just noticed this has to be a valid hostname format (aka no underscores).

Suggest `hostname: ibm-cloud-iks-hostname`

**Comment by agrare** [comment]
This will conflict due to the lack of an object storage manager on the morphy branch, preemptive backport PR :wink:  https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/280

File: app/models/manageiq/providers/ibm_cloud/container_manager/refresh_worker.rb [hunk 15]
```diff
@@ -0,0 +1,8 @@
+class ManageIQ::Providers::IbmCloud::ContainerManager::RefreshWorker < ManageIQ::Providers::BaseManager::RefreshWorker
+  require_nested :Runner
+  require_nested :WatchThread
+
+  def self.settings_name
+    :event_catcher_ibm_cloud_iks
```
Comment (agrare) [inline app/models/manageiq/providers/ibm_cloud/container_manager/refresh_worker.rb:6]: Just noticed this should be `:ems_refresh_worker_ibm_cloud_iks` can you fix in a follow-up @nasark ?

**Comment by Fryguy** [comment]
@nasark A conflict occurred during the backport of this pull request to `morphy`.

If this pull request is based on another pull request that has not been marked for backport, add the appropriate labels to the other pull request. Otherwise, please create a new pull request direct to the `morphy` branch in order to resolve this.

Conflict details:

```diff
diff --cc app/models/manageiq/providers/ibm_cloud/inventory/collector.rb
index a2d29fb,8393da9..0000000
--- a/app/models/manageiq/providers/ibm_cloud/inventory/collector.rb
+++ b/app/models/manageiq/providers/ibm_cloud/inventory/collector.rb
@@@ -1,4 -1,6 +1,9 @@@
  class ManageIQ::Providers::IbmCloud::Inventory::Collector < ManageIQ::Providers::Inventory::Collector
++<<<<<<< HEAD
++=======
+   require_nested :ContainerManager
+   require_nested :ObjectStorage
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
    require_nested :PowerVirtualServers
    require_nested :VPC
  end
diff --cc app/models/manageiq/providers/ibm_cloud/inventory/parser.rb
index 6763419,0839e28..0000000
--- a/app/models/manageiq/providers/ibm_cloud/inventory/parser.rb
+++ b/app/models/manageiq/providers/ibm_cloud/inventory/parser.rb
@@@ -1,4 -1,6 +1,9 @@@
  class ManageIQ::Providers::IbmCloud::Inventory::Parser < ManageIQ::Providers::Inventory::Parser
++<<<<<<< HEAD
++=======
+   require_nested :ContainerManager
+   require_nested :ObjectStorage
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
    require_nested :PowerVirtualServers
    require_nested :VPC
  end
diff --cc app/models/manageiq/providers/ibm_cloud/inventory/persister.rb
index 2adf9a4,32f12eb..0000000
--- a/app/models/manageiq/providers/ibm_cloud/inventory/persister.rb
+++ b/app/models/manageiq/providers/ibm_cloud/inventory/persister.rb
@@@ -1,4 -1,6 +1,9 @@@
  class ManageIQ::Providers::IbmCloud::Inventory::Persister < ManageIQ::Providers::Inventory::Persister
++<<<<<<< HEAD
++=======
+   require_nested :ContainerManager
+   require_nested :ObjectStorage
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
    require_nested :PowerVirtualServers
    require_nested :VPC
  end
diff --cc config/secrets.defaults.yml
index 976b4d6,51a41ff..0000000
--- a/config/secrets.defaults.yml
+++ b/config/secrets.defaults.yml
@@@ -8,3 -8,15 +8,18 @@@ test
      api_key: IBM_CLOUD_VPC_API_KEY
    ibm_cloud_vpc:
      <<: *ibm_cloud_vpc_defaults
++<<<<<<< HEAD
++=======
+   ibm_cloud_object_storage_defaults: &ibm_cloud_object_storage_defaults
+     api_key: IBM_CLOUD_OBJECT_STORAGE_API_KEY
+     access_key: IBM_CLOUD_OBJECT_STORAGE_ACCESS_KEY
+     secret_key: IBM_CLOUD_OBJECT_STORAGE_SECRET_KEY
+   ibm_cloud_object_storage:
+     <<: *ibm_cloud_object_storage_defaults
+   iks_defaults: &iks_defaults
+     api_key: IBM_CLOUD_IKS_API_KEY
+     hostname: ibm-cloud-iks-hostname
+     port: 1234
+   iks:
+     <<: *iks_defaults
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
diff --cc config/settings.yml
index c13c8ea,1df9f76..0000000
--- a/config/settings.yml
+++ b/config/settings.yml
@@@ -66,3 -76,5 +74,8 @@@
        :ems_refresh_worker:
          :ems_refresh_worker_ibm_cloud_power_virtual_servers: {}
          :ems_refresh_worker_ibm_cloud_vpc: {}
++<<<<<<< HEAD
++=======
+         :ems_refresh_worker_ibm_cloud_object_storage: {}
+         :ems_refresh_worker_ibm_cloud_iks: {}
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
diff --cc spec/factories/ext_management_system.rb
index 6111916,270f3d4..0000000
--- a/spec/factories/ext_management_system.rb
+++ b/spec/factories/ext_management_system.rb
@@@ -23,3 -23,33 +23,36 @@@ FactoryBot.define d
            :class  => "ManageIQ::Providers::IbmCloud::VPC::NetworkManager",
            :parent => :ems_cloud
  end
++<<<<<<< HEAD
++=======
+ 
+ FactoryBot.define do
+   factory :ems_ibm_cloud_object_storage_storage,
+           :class  => "ManageIQ::Providers::IbmCloud::ObjectStorage::StorageManager",
+           :parent => :ems_storage
+ end
+ 
+ FactoryBot.define do
+   factory :ems_ibm_cloud_iks, :class => "ManageIQ::Providers::IbmCloud::ContainerManager", :parent => :ems_container do
+     provider_region { "ca-tor" }
+   end
+ 
+   factory :ems_ibm_cloud_iks_with_vcr_authentication, :parent => :ems_ibm_cloud_iks do
+     after(:create) do |ems|
+       api_key = Rails.application.secrets.iks[:api_key]
+ 
+       ems.default_endpoint.update!(
+         :hostname          => Rails.application.secrets.iks[:hostname],
+         :port              => Rails.application.secrets.iks[:port],
+         :security_protocol => "ssl-without-validation"
+       )
+ 
+       ems.authentications << FactoryBot.create(
+         :authentication,
+         :authtype => "bearer",
+         :auth_key => api_key
+       )
+     end
+   end
+ end
++>>>>>>> 81b40c4... Merge pull request #272 from nasark/add_ibmcloud_kubernetes_support
```

**Comment by Fryguy** [comment]
Backported to `morphy` via https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/280
